lib/Sema/TypeCheckAttr.cpp

```diff
@@ -3157,20 +3157,43 @@ DynamicallyReplacedDeclRequest::evaluate(Evaluator &evaluator,
   return nullptr;
 }
 
-/// Returns true if the given type conforms to `Differentiable` in the given
-/// module.
-static bool conformsToDifferentiable(Type type, DeclContext *DC) {
+/// If the given type conforms to `Differentiable` in the given context, returns
+/// the `ProtocolConformanceRef`. Otherwise, returns an invalid
+/// `ProtocolConformanceRef`.
+///
+/// This helper verifies that the `TangentVector` type witness is valid, in case
+/// the conformance has not been fully checked and the type witness cannot be
+/// resolved.
+static ProtocolConformanceRef getDifferentiableConformance(Type type,
+                                                           DeclContext *DC) {
   auto &ctx = type->getASTContext();
   auto *differentiableProto =
       ctx.getProtocol(KnownProtocolKind::Differentiable);
-  auto conf = TypeChecker::conformsToProtocol(
-      type, differentiableProto, DC, ConformanceCheckFlags::InExpression);
+  auto conf =
+      TypeChecker::conformsToProtocol(type, differentiableProto, DC, None);
   if (!conf)
-    return false;
+    return ProtocolConformanceRef();
   // Try to get the `TangentVector` type witness, in case the conformance has
-  // not been fully checked and the type witness cannot be resolved.
+  // not been fully checked.
+  Type tanType = conf.getTypeWitnessByName(type, ctx.Id_TangentVector);
+  if (tanType.isNull() || tanType->hasError())
+    return ProtocolConformanceRef();
+  return conf;
+};
+
+/// Returns true if the given type conforms to `Differentiable` in the given
+/// contxt. If `tangentVectorEqualsSelf` is true, also check whether the given
+/// type satisfies `TangentVector == Self`.
+static bool conformsToDifferentiable(Type type, DeclContext *DC,
```

Comment:
A general comment is that the clients of this function are not consistent about whether they're passing interface types or contextual types to this function. The conformance check will most certainly fail if the interface-type-passing callers are not amended. I think you should try to just accept the interface type here, assert as much as a precondition, and map them into the passed-in declaration context yourself.

Reply:
This will also keep you from using contextual types in diagnostics.

Reply:
Thank you for the review! I'll try to refine `getDifferentiableConformance` to accept only interface types.

Reply:
Is my understanding correct?
- `TypeChecker::conformsToProtocol` should be called on contextual types. It doesn't (always) work on interface types?
- Diagnostics require interface types. Contextual types may show up as `τ_0_0`, which is bad.
- Your suggestion: it would be nice for this `conformsToDifferentiable` helper function to be consistent in accepting interface types or contextual types.

---

Context: this helper function is used for type-checking `@differentiable` and `@derivative` attributes. `@differentiable` attributes may have an associated `where` clause. which is resolved to a `GenericSignature`/`GenericEnvironment`:

```swift
@differentiable(wrt: (x) where T: Differentiable)
func identity<T>(_ x: T) -> T { x }
```

We want to verify that the types of all `wrt` parameters and the function result type have well-formed conformances to the `Differentiable` protocol.

There is no `DeclContext` for `where` clauses, hence why some callers of `conformsToDifferentiable` first map types into the `GenericEnvironment` before passing them to the function:

https://github.com/apple/swift/blob/dcfc655491f7c1666c52e2bed7db17691cdf1b8f/lib/Sema/TypeCheckAttr.cpp#L3196-L3204

So I'm not sure it's possible to limit `conformsToDifferentiable` to accept only interface types and a `DeclContext`.

It's also not possible for `conformsToDifferentiable` to accept only interface types and a `GenericEnvironment`, because a `DeclContext` is necessary to call `TypeCheckProtocol::conformsToProtocol`.

---

We could make `conformsToDifferentiable` accept only interface types, taking both a `DeclContext` and an optional `GenericEnvironment` (defaulting to `DeclContext::getGenericEnvironmentOfContext()`). I can explore this now, since I can't think of other solutions.

Any suggestions welcome!

Reply:
Seems like we're at an impasse because of the modeling choices then. I won't hold up this patch for now, but we should discuss refactoring this code in slack.